solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,377 @@
+    ///           and output index has been revealed to the Bridge.
+    ///         - The deposit has not been swept yet.
+    ///         - The deposit is targeted into the TBTCVault.
+    ///         - The optimistic minting is not paused.
+    ///         After calling this function, the Minter has to wait for
+    ///         OPTIMISTIC_MINTING_DELAY before finalizing the mint with a call
+    ///         to finalizeOptimisticMint.
+    /// @dev The deposit done on the Bitcoin side must be revealed early enough
+    ///      to the Bridge on Ethereum to pass the Bridge's validation. The
+    ///      validation passes successfully only if the deposit reveal is done
+    ///      respectively earlier than the moment when the deposit refund
+    ///      locktime is reached, i.e. the deposit becomes refundable. It may
+    ///      happen that the wallet does not sweep a revealed deposit and one of
+    ///      the Minters requests an optimistic mint for that deposit just
+    ///      before the locktime is reached. Guardians must cancel optimistic
+    ///      minting for this deposit because the wallet will not be able to
+    ///      sweep it. The on-chain optimistic minting code does not perform any
+    ///      validation for gas efficiency: it would have to perform the same
+    ///      validation as `validateDepositRefundLocktime` and expect the entire
+    ///      `DepositRevealInfo` to be passed to assemble the expected script
+    ///      hash on-chain. Guardians must validate if the deposit happened on
+    ///      Bitcoin, that the script hash has the expected format, and that the
+    ///      wallet is an active one so they can also validate the time left for
+    ///      the refund.
+    function optimisticMint(bytes32 fundingTxHash, uint32 fundingOutputIndex)
+        external
+        onlyMinter
+        whenOptimisticMintingNotPaused
+    {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
+
+        require(deposit.revealedAt != 0, "The deposit has not been revealed");
+        require(deposit.sweptAt == 0, "The deposit is already swept");
+        require(deposit.vault == address(this), "Unexpected vault address");
+
+        /* solhint-disable-next-line not-rely-on-time */
+        pendingOptimisticMints[depositKey] = block.timestamp;
+
+        emit OptimisticMintingRequested(
+            msg.sender,
+            deposit.depositor,
+            deposit.amount,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    /// @notice Allows a Minter to finalize previously requested optimistic
+    ///         minting. The following conditions must be met:
+    ///         - The optimistic minting has been requested for the given
+    ///           deposit.
+    ///         - The deposit has not been swept yet.
```

Comment:
Let's separate how this will work in January and how this will - potentially - work a couple of months later.

In January, wallets are not yet sweeping. The 3h delay is the time for the deposit transaction to be revealed on Ethereum, mined on Bitcoin, and become sufficiently confirmed on both Ethereum and Bitcoin. (20 + 6 confirmations?).

Once we activate sweeping, we will have to decide if the optimistic minting gets paused or if it stays as a quickminting solution. If we pause it, we have no problem here. If we use it as a quickminting solution, we will have kind of a race between the optimistic minting and wallets sweeping.

If the optimistic minting gets requested, and the wallet sweeps first, the finalization will not be possible. The request may stay dangling or potentially be canceled by Guardians to clean up the optimistic minting state. I am not sure if it is worth the gas expenditure though.

If the optimistic minting gets requested, and it gets finalized before the wallet sweeps, the `optimisticMintingDebt` will get incurred. When the wallet finalizes the sweep of the deposit, it will be partially paid off. This will also be a scenario when we activate the sweeping of deposits from January. Please see the unit tests added in 6d5eddeaa3f63bbf1fd4d5ba5a7d02ef572d0b38 - I worked on them while you were reviewing.